README.md
# timeproxy
[![npm version](https://badge.fury.io/js/timeproxy.svg)](https://badge.fury.io/js/timeproxy) [![Build Status](https://travis-ci.org/selbekk/timeproxy.svg?branch=master)](https://travis-ci.org/selbekk/timeproxy) [![Coverage Status](https://coveralls.io/repos/github/selbekk/timeproxy/badge.svg?branch=master)](https://coveralls.io/github/selbekk/timeproxy?branch=master) [![Y2K Compliant](https://img.shields.io/badge/%F0%9F%97%93-Y2K%20Compliant-bd6867.svg)]()


> Simple library for time constants

**Now with template literal support**

Creating time constants can be a pain. You want to name them
after what they are, like `SESSION_TIMEOUT` or `MAX_REQUESTS`,
but you also want to make sure the code reflects how long of
a period it is. Meet `timeproxy`.

Using some quite clever ES2015 proxies, you can now have both.

## Usage

```bash
npm install timeproxy
```

`timeproxy` works by parsing the name of the "constant" you
specify and returning the amount of milliseconds you require. 

```js
import tp from 'timeproxy';

const TIMEOUT_LIMIT = tp.FIVE_SECONDS;
const AGE_LIMIT = tp.ONE_WEEK_AND_SIX_DAYS;
```

If you prefer, you can use template literals instead:

```js
import tp from 'timeproxy';

const TIMEOUT_LIMIT = tp`five seconds`;
const AGE_LIMIT = tp`1 week and 6 days`;
const SESSION_TIMEOUT = tp`${60} minutes`;
```

There's support for seconds, minutes, hours, days and weeks. You can even write 
fractions!

Here's a few examples of what you can do: 

```js
import tp from 'timeproxy';

const REQUEST_TIMEOUT = tp.THIRTY_SECONDS;
const UPDATE_DELAY = tp.HALF_A_SECOND;
const COOKIE_EXPIRATION = tp.FOUR_WEEKS;
const IN_ALMOST_A_MINUTE = tp.IN_FIFTY_NINE_SECONDS;

const REQUEST_TIMEOUT = tp`thirty seconds`;
const UPDATE_DELAY = tp`.5 seconds`;
const COOKIE_EXPIRATION = tp`4 weeks`;
const IN_ALMOST_A_MINUTE = tp`IN_59-SEconds`;
```

I've [written a blog post](https://medium.com/@selbekk/timeproxy-a-library-for-making-readable-time-constants-5f6bdd9c598d) 
about this as well - please refer to it for more examples. 

## Contribute!

If there are features you'd like to see in this tiny library,
please let me know through an issue. If you feel up for it,
please create a pull request and I'll make sure to look at it
as soon as possible.


CHANGELOG.md
# Changelog

## v1.3.0

- Add support for months, years and decades
- Add TypeScript support (for functions only)

## v1.2.1

- Return 0 when called with either no or a falsy argument

## v1.2.0

- Added support for using `timeproxy` as a tagged template literal as well!

## v1.1.0

- Added support for "in" prefix and "ago" suffix for relative time
- Added support for "a" and "an" instead of "one"
- Added support for "of" as a separator
- Added support for halves, thirds and quarters
- Updated readme

## v1.0.7

- First working version, because reasons

## v1.0.1 - v1.0.6

- All broken. A wasteland. 🗑

## v1.0.0

- Initial version 💪


src/index.test.js
import tp from ".";

const MILLISECOND = 1;
const SECOND = 1000 * MILLISECOND;
const MINUTE = 60 * SECOND;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;
const WEEK = 7 * DAY;
const MONTH = 30 * DAY;
const YEAR = 365 * DAY;
const DECADE = 10 * YEAR;

describe("properties without units", () => {
  it("returns correct with just numbers", () => {
    expect(tp.ZERO).toBe(0);
    expect(tp.ONE).toBe(1);
    expect(tp.NINE).toBe(9);
    expect(tp.THIRTEEN).toBe(13);
  });

  it("throws an error if number is unparseable", () => {
    expect(() => tp.SOME_UNKNOWN_VARIABLE).toThrow(
      /Could not parse expression/
    );
  });
});

describe("properties with units", () => {
  it("returns correct with milliseconds", () => {
    expect(tp.ONE_MILLISECOND).toBe(1 * MILLISECOND);
    expect(tp.NINE_HUNDRED_MILLISECONDS).toBe(900 * MILLISECOND);
    expect(tp.NINETY_THOUSAND_MILLISECONDS).toBe(90000 * MILLISECOND);
  });

  it("returns correct with seconds", () => {
    expect(tp.TWO_SECONDS).toBe(2 * SECOND);
    expect(tp.THIRTY_SECONDS).toBe(30 * SECOND);
    expect(tp.HUNDRED_SECONDS).toBe(100 * SECOND);
  });

  it("returns correct with minutes", () => {
    expect(tp.THREE_MINUTES).toBe(3 * MINUTE);
    expect(tp.FIFTEEN_MINUTES).toBe(15 * MINUTE);
    expect(tp.HUNDRED_MINUTES).toBe(100 * MINUTE);
  });

  it("returns correct with hours", () => {
    expect(tp.ONE_HOURS).toBe(1 * HOUR);
    expect(tp.FOUR_HOURS).toBe(4 * HOUR);
    expect(tp.FOURTY_HOURS).toBe(40 * HOUR);
  });

  it("returns correct with days", () => {
    expect(tp.SEVEN_DAYS).toBe(7 * DAY);
    expect(tp.SEVENTEEN_DAYS).toBe(17 * DAY);
    expect(tp.EIGHTY_DAYS).toBe(80 * DAY);
  });

  it("returns correct with weeks", () => {
    expect(tp.FIVE_WEEKS).toBe(5 * WEEK);
    expect(tp.FIFTEEN_WEEKS).toBe(15 * WEEK);
    expect(tp.FIFTY_WEEKS).toBe(50 * WEEK);
  });

  it("returns correct with months", () => {
    expect(tp.ONE_MONTH).toBe(1 * MONTH);
    expect(tp.FIFTEEN_MONTHS).toBe(15 * MONTH);
    expect(tp.FIFTY_MONTHS).toBe(50 * MONTH);
  });

  it("returns correct with years", () => {
    expect(tp.ONE_YEAR).toBe(1 * YEAR);
    expect(tp.FIFTEEN_YEARS).toBe(15 * YEAR);
    expect(tp.FIFTY_YEARS).toBe(50 * YEAR);
  });

  it("returns correct with decades", () => {
    expect(tp.ONE_DECADE).toBe(1 * DECADE);
    expect(tp.FIFTEEN_DECADES).toBe(15 * DECADE);
    expect(tp.FIFTY_DECADES).toBe(50 * DECADE);
  });
});

describe("properties with combined numbers", () => {
  it("works with two numbers", () => {
    expect(tp.TWENTY_FIVE_SECONDS).toBe(25 * SECOND);
  });

  it("works with several numbers", () => {
    expect(tp.TWO_HUNDRED_TWENTY_FIVE_SECONDS).toBe(225 * SECOND);
  });

  it("ignores the AND term", () => {
    expect(tp.HUNDRED_AND_THIRTY_TWO).toBe(132);
    expect(tp.HUNDRED_AND_THIRTY_TWO_SECONDS).toBe(132 * SECOND);
  });
});

describe("properties with combined units", () => {
  it("works with two units", () => {
    expect(tp.TWO_MINUTES_AND_THREE_SECONDS).toBe(2 * MINUTE + 3 * SECOND);
    expect(tp.NINETY_HOURS_FOUR).toBe(90 * HOUR + 4);
  });

  it("works witn several units", () => {
    expect(tp.ONE_HOURS_TWO_MINUTES_THREE_SECONDS).toBe(
      HOUR + 2 * MINUTE + 3 * SECOND
    );
    expect(tp.NINE_WEEKS_TWELVE_DAYS_SEVEN_MINUTES_AND_HALF_A_SECOND).toBe(
      WEEK * 9 + DAY * 12 + MINUTE * 7 + 500
    );
  });
});

describe("properties with single unit", () => {
  it("works with single unit", () => {
    expect(tp.ONE_SECOND).toBe(SECOND);
    expect(tp.ONE_MINUTE).toBe(MINUTE);
    expect(tp.ONE_HOUR).toBe(HOUR);
    expect(tp.ONE_DAY).toBe(DAY);
    expect(tp.ONE_WEEK).toBe(WEEK);
    expect(tp.ONE_MONTH).toBe(MONTH);
    expect(tp.ONE_YEAR).toBe(YEAR);
    expect(tp.ONE_DECADE).toBe(DECADE);
  });
});

describe("properties with fraction units", () => {
  it("works with halves", () => {
    expect(tp.ONE_HALF_SECOND).toBe(500);
    expect(tp.HALF_A_SECOND).toBe(500);
    expect(tp.THREE_HALVES).toBe(1.5);
  });
  it("works with thirds", () => {
    expect(tp.A_THIRD_DAY).toBe(HOUR * 8);
    expect(tp.A_THIRD).toBe(1 / 3);
  });
  it("works with quarters", () => {
    expect(tp.THREE_QUARTERS_HOUR).toBe((HOUR / 4) * 3);
    expect(tp.A_QUARTER_DAY).toBe(HOUR * 6);
  });
});

describe("properties with a and an", () => {
  it("counts for one", () => {
    expect(tp.A_SECOND).toBe(SECOND);
    expect(tp.AN_HOUR).toBe(HOUR);
  });
  it("doesnt judge you for not knowing grammer", () => {
    expect(tp.AN_SECOND).toBe(SECOND);
    expect(tp.A_HOUR).toBe(HOUR);
  });
  it("see what I did there?", () => {});
});

describe("properties with of", () => {
  it("reads better, but doesn't cause trouble", () => {
    expect(tp.HALF_OF_A_MINUTE).toBe(MINUTE / 2);
    expect(tp.THIRTY_OF_SECONDS).toBe(SECOND * 30);
    expect(tp.A_QUARTER_OF_AN_HOUR).toBe(MINUTE * 15);
  });
});

describe("properties that are relative in time", () => {
  it("works with past", () => {
    const now = Date.now();
    Date.now = jest.fn(() => now);

    expect(tp.AN_HOUR_AGO).toBe(now - HOUR);
    expect(tp.A_MINUTE_THIRTY_SECONDS_AGO).toBe(now - (MINUTE + 30 * SECOND));
  });
  it("works with future", () => {
    const now = Date.now();
    Date.now = jest.fn(() => now);

    expect(tp.IN_AN_HOUR).toBe(now + HOUR);
    expect(tp.IN_A_WEEK_AND_SIX_DAYS).toBe(now + WEEK + DAY * 6);
  });
});

describe("edge cases", () => {
  it("works with any casing", () => {
    expect(tp.one_second).toBe(SECOND);
    expect(tp.one_Week).toBe(WEEK);
    expect(tp.ONE_hour).toBe(HOUR);
    expect(tp.OnE_DaY).toBe(DAY);
  });

  it("works with several multipliers back to back", () => {
    expect(tp.HUNDRED_THOUSAND_SECONDS).toBe(100 * 1000 * SECOND);
  });
});

describe("tagged template literal", () => {
  it("works with underscores", () => {
    expect(tp`one_second`).toBe(SECOND);
    expect(tp`ONE_WEEK`).toBe(WEEK);
  });
  it("works with spaces", () => {
    expect(tp`one second`).toBe(SECOND);
    expect(tp`ONE WEEK`).toBe(WEEK);
  });
  it("works with hyphens", () => {
    expect(tp`one-second`).toBe(SECOND);
    expect(tp`ONE-WEEK`).toBe(WEEK);
  });
  it("works with a mix of approaches", () => {
    expect(tp`ONE-WEEK_AND seven minutes`).toBe(WEEK + MINUTE * 7);
  });

  it("works with arguments", () => {
    expect(tp`1 day ${60} minutes and ${30} seconds`).toBe(
      DAY + MINUTE * 60 + 30 * SECOND
    );
  });
});

describe("numerics", () => {
  it("works with proxies", () => {
    expect(tp.A_MINUTE_20_SECONDS).toBe(MINUTE + 20 * 1000);
  });
  it("works with tagged template literals", () => {
    expect(tp`a minute 60 seconds`).toBe(2 * MINUTE);
    expect(tp`1 minute 60 seconds`).toBe(2 * MINUTE);
    expect(tp`1 minute .5 seconds`).toBe(MINUTE + 500);
  });
});

describe("alternative usage", () => {
  it("works when being called as a function", () => {
    expect(tp()).toBe(0);
    expect(tp(`a minute 45`)).toBe(MINUTE + 45);
    expect(tp(`a minute ${45}`)).toBe(MINUTE + 45);
  });
});


__tests__/index.test.js
import tp from '../src';

const MILLISECOND = 1;
const SECOND = 1000 * MILLISECOND;
const MINUTE = 60 * SECOND;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;
const WEEK = 7 * DAY;

describe('properties without units', () => {
    it('returns correct with just numbers', () => {
        expect(tp.ZERO).toBe(0);
        expect(tp.ONE).toBe(1);
        expect(tp.NINE).toBe(9);
        expect(tp.THIRTEEN).toBe(13);
    });

    it('throws an error if number is unparseable', () => {
        expect(() => tp.SOME_UNKNOWN_VARIABLE)
            .toThrow(/Could not parse expression/);
    });
});

describe('properties with units', () => {
    it('returns correct with milliseconds', () => {
        expect(tp.ONE_MILLISECOND).toBe(1 * MILLISECOND);
        expect(tp.NINE_HUNDRED_MILLISECONDS).toBe(900 * MILLISECOND);
        expect(tp.NINETY_THOUSAND_MILLISECONDS).toBe(90000 * MILLISECOND);
    });

    it('returns correct with seconds', () => {
        expect(tp.TWO_SECONDS).toBe(2 * SECOND);
        expect(tp.THIRTY_SECONDS).toBe(30 * SECOND);
        expect(tp.HUNDRED_SECONDS).toBe(100 * SECOND);
    });

    it('returns correct with minutes', () => {
        expect(tp.THREE_MINUTES).toBe(3 * MINUTE);
        expect(tp.FIFTEEN_MINUTES).toBe(15 * MINUTE);
        expect(tp.HUNDRED_MINUTES).toBe(100 * MINUTE);
    });

    it('returns correct with hours', () => {
        expect(tp.ONE_HOURS).toBe(1 * HOUR);
        expect(tp.FOUR_HOURS).toBe(4 * HOUR);
        expect(tp.FOURTY_HOURS).toBe(40 * HOUR);
    });

    it('returns correct with days', () => {
        expect(tp.SEVEN_DAYS).toBe(7 * DAY);
        expect(tp.SEVENTEEN_DAYS).toBe(17 * DAY);
        expect(tp.EIGHTY_DAYS).toBe(80 * DAY);
    });

    it('returns correct with weeks', () => {
        expect(tp.FIVE_WEEKS).toBe(5 * WEEK);
        expect(tp.FIFTEEN_WEEKS).toBe(15 * WEEK);
        expect(tp.FIFTY_WEEKS).toBe(50 * WEEK);
    });
});

describe('properties with combined numbers', () => {
    it('works with two numbers', () => {
        expect(tp.TWENTY_FIVE_SECONDS).toBe(25 * SECOND);
    });

    it('works with several numbers', () => {
        expect(tp.TWO_HUNDRED_TWENTY_FIVE_SECONDS).toBe(225 * SECOND);
    });

    it('ignores the AND term', () => {
        expect(tp.HUNDRED_AND_THIRTY_TWO).toBe(132);
        expect(tp.HUNDRED_AND_THIRTY_TWO_SECONDS).toBe(132 * SECOND);
    });
});

describe('properties with combined units', () => {
    it('works with two units', () => {
        expect(tp.TWO_MINUTES_AND_THREE_SECONDS).toBe(2 * MINUTE + 3 * SECOND);
        expect(tp.NINETY_HOURS_FOUR).toBe(90 * HOUR + 4);
    });

    it('works witn several units', () => {
        expect(tp.ONE_HOURS_TWO_MINUTES_THREE_SECONDS).toBe(HOUR + 2 * MINUTE + 3 * SECOND);
        expect(tp.NINE_WEEKS_TWELVE_DAYS_SEVEN_MINUTES_AND_HALF_A_SECOND).toBe(WEEK * 9 + DAY * 12 + MINUTE * 7 + 500);
    });
});

describe('properties with single unit', () => {
    it('works with single unit', () => {
        expect(tp.ONE_SECOND).toBe(SECOND);
        expect(tp.ONE_MINUTE).toBe(MINUTE);
        expect(tp.ONE_HOUR).toBe(HOUR);
        expect(tp.ONE_DAY).toBe(DAY);
        expect(tp.ONE_WEEK).toBe(WEEK);
    });
});

describe('properties with fraction units', () => {
    it('works with halves', () => {
        expect(tp.ONE_HALF_SECOND).toBe(500);
        expect(tp.HALF_A_SECOND).toBe(500);
        expect(tp.THREE_HALVES).toBe(1.5);
    });
    it('works with thirds', () => {
        expect(tp.A_THIRD_DAY).toBe(HOUR * 8);
        expect(tp.A_THIRD).toBe(1/3);
    });
    it('works with quarters', () => {
        expect(tp.THREE_QUARTERS_HOUR).toBe(HOUR / 4 * 3);
        expect(tp.A_QUARTER_DAY).toBe(HOUR * 6);
    });
});

describe('properties with a and an', () => {
    it('counts for one', () => {
        expect(tp.A_SECOND).toBe(SECOND);
        expect(tp.AN_HOUR).toBe(HOUR);
    });
    it('doesnt judge you for not knowing grammer', () => {
        expect(tp.AN_SECOND).toBe(SECOND);
        expect(tp.A_HOUR).toBe(HOUR);
    });
    it('see what I did there?', () => {});
});

describe('properties with of', () => {
    it('reads better, but doesn\'t cause trouble', () => {
        expect(tp.HALF_OF_A_MINUTE).toBe(MINUTE / 2);
        expect(tp.THIRTY_OF_SECONDS).toBe(SECOND * 30);
        expect(tp.A_QUARTER_OF_AN_HOUR).toBe(MINUTE * 15);
    });
});

describe('properties that are relative in time', () => {
    it('works with past', () => {
        const now = Date.now();
        Date.now = jest.fn(() => now);

        expect(tp.AN_HOUR_AGO).toBe(now - HOUR);
        expect(tp.A_MINUTE_THIRTY_SECONDS_AGO).toBe(now - (MINUTE + 30 * SECOND));
    });
    it('works with future', () => {
        const now = Date.now();
        Date.now = jest.fn(() => now);

        expect(tp.IN_AN_HOUR).toBe(now + HOUR);
        expect(tp.IN_A_WEEK_AND_SIX_DAYS).toBe(now + (WEEK) + (DAY * 6));
    });
});

describe('edge cases', () => {
    it('works with any case', () => {
        expect(tp.one_second).toBe(SECOND);
        expect(tp.one_Week).toBe(WEEK);
        expect(tp.ONE_hour).toBe(HOUR);
        expect(tp.OnE_DaY).toBe(DAY);
    });

    it('works with several multipliers back to back', () => {
        expect(tp.HUNDRED_THOUSAND_SECONDS).toBe(100 * 1000 * SECOND);
    });
});

describe('tagged template literal', () => {
    it('works with underscores', () => {
        expect(tp`one_second`).toBe(SECOND);
        expect(tp`ONE_WEEK`).toBe(WEEK);
    });
    it('works with spaces', () => {
        expect(tp`one second`).toBe(SECOND);
        expect(tp`ONE WEEK`).toBe(WEEK);
    });
    it('works with hyphens', () => {
        expect(tp`one-second`).toBe(SECOND);
        expect(tp`ONE-WEEK`).toBe(WEEK);
    });
    it('works with a mix of approaches', () => {
        expect(tp`ONE-WEEK_AND seven minutes`).toBe(WEEK + (MINUTE * 7));
    });

    it('works with arguments', () => {
        const sixty = 60;
        expect(tp`1 day ${60} minutes and ${30} seconds`)
            .toBe(DAY + (MINUTE * 60) + (30 * SECOND));
    });
});

describe('numerics', () => {
    it('works with proxies', () => {
        expect(tp.A_MINUTE_20_SECONDS).toBe(MINUTE + (20 * 1000));
    });
    it('works with tagged template literals', () => {
        expect(tp`a minute 60 seconds`).toBe(2 * MINUTE);
        expect(tp`1 minute 60 seconds`).toBe(2 * MINUTE);
        expect(tp`1 minute .5 seconds`).toBe(MINUTE + 500);
    });
});

describe('alternative usage', () => {
    it('works when being called as a function', () => {
        expect(tp()).toBe(0);
        expect(tp(`a minute 45`)).toBe(MINUTE + 45);
        expect(tp(`a minute ${45}`)).toBe(MINUTE + 45);
    });
});


src/index.js
import { numbers, multipliers, units } from './constants';

export default new Proxy(() => {}, {
    apply(a, b, args) {
        if (!args[0] || !args[0].raw) {
            return find(...args);
        }
        return find(String.raw(...args));
    },
    get(target, name) {
        return find(name);
    }
});

const isNumeric = token => !isNaN(token);
const isNumber = token => numbers[token] !== undefined;
const isMultiplier = token => multipliers[token] !== undefined;
const isUnit = token => units[token] !== undefined;

function parse(token) {
    if (isNumeric(token)) {
        return { type: 'number', value: Number(token) };
    }
    if (isNumber(token)) {
        return { type: 'number', value: numbers[token] };
    }
    if (isMultiplier(token)) {
        return { type: 'multiplier', value: multipliers[token] };
    }
    if (isUnit(token)) {
        return { type: 'unit', value: units[token] };
    }
    throw new Error(`Could not parse expression ${token}.`);
}

function find(name) {
    if(!name) {
        return 0;
    }

    const isFuture = name.startsWith('IN');
    const isPast = name.endsWith('AGO');

    const terms = name.split(/[_\s\-]+/)
        .map(term => term.toUpperCase())
        .filter(term => !['IN', 'AGO', 'AND', 'OF'].includes(term))
        .map(parse);

    let sum = 0;
    let partSum = 0;

    for (let i = 0; i < terms.length; i++) {
        const term = terms[i];

        switch (term.type) {
            case 'number': {
                partSum += term.value;
                break;
            }
            case 'multiplier': {
                partSum = (partSum || 1) * term.value;
                break;
            }
            case 'unit': {
                sum = sum + partSum * term.value;
                partSum = 0;
                break;
            }
        }
    }

    sum += partSum;

    if (isFuture) {
        return Date.now() + sum;
    } else if (isPast) {
        return Date.now() - sum;
    } 
    return sum;
}


types/types.d.ts
import { multipliers, numbers, units } from "../src/constants";
declare module "timeproxy" {
  type ValidNumber = keyof typeof numbers | number;
  type ValidMultiplier = keyof typeof multipliers;
  type ValidUnit = keyof typeof units;

  type ValidUnitWithMultiplier<T extends string> = T extends `${
    | ValidMultiplier
    | ValidNumber} ${infer Rest}`
    ? ValidUnitWithMultiplier<Rest>
    : T extends ValidUnit // needs to end with unit
    ? true
    : false;

  type ValidNumberWithUnit = `${ValidNumber} ${ValidUnit}`;

  type ValidValue<T extends string> = ValidUnitWithMultiplier<T> extends true
    ? true
    : T extends ValidNumberWithUnit
    ? true
    : T extends ValidNumber
    ? true
    : false;

  type ValidSentenceInner<T extends string> =
    Uppercase<T> extends `${infer Curr} AND ${infer Rest}`
      ? ValidValue<Uppercase<Curr>> extends true
        ? ValidSentenceInner<Uppercase<Rest>>
        : false
      : ValidValue<Uppercase<T>>; // ingen 'and'

  type ValidSentence<T extends string> = Uppercase<T> extends `IN ${infer Rest}`
    ? ValidSentenceInner<Rest>
    : Uppercase<T> extends `${infer Rest} AGO`
    ? ValidSentenceInner<Rest>
    : ValidSentenceInner<T>;

  type ValidArg<T extends string> = ValidSentence<T> extends true ? T : never;

  const timeproxy = <T extends string>(time: ValidArg<T>): number => {
    return 0;
  };
  export default timeproxy;
}


src/constants.js
export const multipliers = {
  QUARTER: 1 / 4,
  QUARTERS: 1 / 4,
  THIRD: 1 / 3,
  THIRDS: 1 / 3,
  HALF: 1 / 2,
  HALVES: 1 / 2,
  A: 1,
  AN: 1,
  HUNDRED: 100,
  THOUSAND: 1000,
  MILLION: 1000000,
  BILLION: 1000000000,
};

export const numbers = {
  ZERO: 0,
  ONE: 1,
  TWO: 2,
  THREE: 3,
  FOUR: 4,
  FIVE: 5,
  SIX: 6,
  SEVEN: 7,
  EIGHT: 8,
  NINE: 9,
  TEN: 10,
  ELEVEN: 11,
  TWELVE: 12,
  THIRTEEN: 13,
  FOURTEEN: 14,
  FIFTEEN: 15,
  SIXTEEN: 16,
  SEVENTEEN: 17,
  EIGHTEEN: 18,
  NINETEEN: 19,
  TWENTY: 20,
  THIRTY: 30,
  FOURTY: 40,
  FIFTY: 50,
  SIXTY: 60,
  SEVENTY: 70,
  EIGHTY: 80,
  NINETY: 90,
};

export const units = {
  MILLISECOND: 1,
  MILLISECONDS: 1,
  SECOND: 1000,
  SECONDS: 1000,
  MINUTE: 1000 * 60,
  MINUTES: 1000 * 60,
  HOUR: 1000 * 60 * 60,
  HOURS: 1000 * 60 * 60,
  DAY: 1000 * 60 * 60 * 24,
  DAYS: 1000 * 60 * 60 * 24,
  WEEK: 1000 * 60 * 60 * 24 * 7,
  WEEKS: 1000 * 60 * 60 * 24 * 7,
  MONTH: 1000 * 60 * 60 * 24 * 30,
  MONTHS: 1000 * 60 * 60 * 24 * 30,
  YEAR: 1000 * 60 * 60 * 24 * 365,
  YEARS: 1000 * 60 * 60 * 24 * 365,
  DECADE: 1000 * 60 * 60 * 24 * 365 * 10,
  DECADES: 1000 * 60 * 60 * 24 * 365 * 10,
};
